File: backend/python/app/api/timeline.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.core.timeline_event_type import TimelineEventType, TimelinePeriod
from app.db.session import get_db
from app.schemas.timeline import PaginatedTimelineResponse, TimelineListQuery
from app.services import timeline_service

router = APIRouter(dependencies=[Depends(get_current_user)])


@router.get("", response_model=PaginatedTimelineResponse)
def list_timeline_events(
    db: Session = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=100),
    search: str | None = None,
    type: TimelineEventType | None = None,
    service_id: int | None = Query(default=None, gt=0),
    incident_id: int | None = Query(default=None, gt=0),
    alert_id: int | None = Query(default=None, gt=0),
    period: TimelinePeriod = TimelinePeriod.all,
):
    query = TimelineListQuery(
        page=page,
        page_size=page_size,
        search=search,
        type=type,
        service_id=service_id,
        incident_id=incident_id,
        alert_id=alert_id,
        period=period,
    )
    return timeline_service.list_timeline_events(db, query)

File: frontend/src/lib/api/timeline.ts
import type { PaginatedTimelineEvents, TimelineListParams } from '../../types/api/timeline.ts'
import { apiRequest } from './client.ts'

function buildTimelineQuery(params: TimelineListParams = {}): string {
  const searchParams = new URLSearchParams()

  if (params.page !== undefined) {
    searchParams.set('page', String(params.page))
  }
  if (params.page_size !== undefined) {
    searchParams.set('page_size', String(params.page_size))
  }
  if (params.search !== undefined && params.search.length > 0) {
    searchParams.set('search', params.search)
  }
  if (params.type !== undefined) {
    searchParams.set('type', params.type)
  }
  if (params.service_id !== undefined) {
    searchParams.set('service_id', String(params.service_id))
  }
  if (params.incident_id !== undefined) {
    searchParams.set('incident_id', String(params.incident_id))
  }
  if (params.alert_id !== undefined) {
    searchParams.set('alert_id', String(params.alert_id))
  }
  if (params.period !== undefined) {
    searchParams.set('period', params.period)
  }

  const query = searchParams.toString()
  return query.length > 0 ? `?${query}` : ''
}

export function fetchTimeline(params: TimelineListParams = {}): Promise<PaginatedTimelineEvents> {
  return apiRequest<PaginatedTimelineEvents>(`/timeline${buildTimelineQuery(params)}`)
}
